sycl/plugins/level_zero/pi_level0.cpp

```diff
@@ -366,31 +365,85 @@ pi_result _pi_device::initialize() {
   return PI_SUCCESS;
 }
 
-// Crate a new command list to be used in a PI call
+// Retrieve an available command list to be used in a PI call
 pi_result
-_pi_device::createCommandList(ze_command_list_handle_t *ZeCommandList) {
-  // Create the command list, because in L0 commands are added to
+_pi_device::getAvailableCommandList(ze_command_list_handle_t *ZeCommandList,
+                                    ze_command_queue_handle_t ZeCommandQueue,
+                                    ze_fence_handle_t *ZeFence) {
+  // Create/Reuse the command list, because in L0 commands are added to
   // the command lists, and later are then added to the command queue.
-  //
-  // TODO: Figure out how to lower the overhead of creating a new list
-  // for each PI command, if that appears to be important.
+  // Each command list is paired with an associated fence to track when the
+  // command list is available for reuse.
+  _pi_result pi_result = PI_OUT_OF_RESOURCES;
   ze_command_list_desc_t ZeCommandListDesc = {};
   ZeCommandListDesc.version = ZE_COMMAND_LIST_DESC_VERSION_CURRENT;
-
-  // TODO: can we just reset the command-list created when an earlier
-  // command was submitted to the queue?
-  ZE_CALL(zeCommandListCreate(ZeDevice, &ZeCommandListDesc, ZeCommandList));
-
-  return PI_SUCCESS;
+  ze_fence_desc_t ZeFenceDesc = {};
+
+  // Initally, we need to check if a command list has already been created
+  // on this device that is available for use. If so, then reuse that
+  // L0 Command List and Fence for this PI call.
+  if (this->ZeAvailableCommandListCache.size() > 0) {
+    this->ZeCommandListCacheMutex.lock();
+    *ZeCommandList = this->ZeAvailableCommandListCache.front();
+    *ZeFence = this->ZeCommandListFenceMap[*ZeCommandList];
+    this->ZeAvailableCommandListCache.pop_front();
+    this->ZeCommandListCacheMutex.unlock();
+    return PI_SUCCESS;
+  } else {
```

Comment:
Else branch is unnecessary here.
https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return

Reply:
Else removed, thanks you for your feedback! Please let me know if there is anything else missed!